# **异常处理**

在‘3.5.程序执行’中已经详细说明了Larva的异常机制，本节主要说明异常相关的内建语法元素

## **异常的抛出**

Larva的异常类是指一个实现了`Throwable`接口的类，一般所说的“异常”是指异常类的对象实例

`Throwable`接口的定义很简单，实际上就是继承了`Stringable`接口：
```
public interface Throwable
{
    /*
    返回异常对象包含的信息，或异常本身的描述
    */
    public String str();
}
```

异常的抛出可以使用内建函数`throw`：
```
/*
抛出异常
*/
public void throw(Throwable t)
```
`throw`函数将设置其栈帧的异常为传入的参数`t`，并将其上抛，从逻辑上也可认为是立即中断当前执行并抛出异常，由于异常在`defer`的时候才捕获，
因此当前函数中`throw`之后的代码不会再被执行

*Note*：
* 由于很多类都实现了`Stringable`接口，这也意味着这些类的对象实例都能作为异常抛出，但实际上它们并不一定是一个异常，或适合作为一个异常
* 虽然字符串本身也实现了`Throwable`接口，将其直接作为异常（信息）抛出也是比较合乎逻辑，但这个异常类型太泛了（只是一个信息），一般也不推荐使用

## **异常的捕获**

### **Catched类**

`Catched`是一个泛型类，是异常对象+traceback的封装，Larva在捕获异常时会捕获到这个类的对象实例

```
/*
封装异常对象+traceback
T必须是一个实现了Throwable接口的类
*/
public class Catched<T>
{
    /*
    重新抛出此异常，需要注意和“throw(c.throwed())”这种用法的区别：
    - rethrow方法会原样上抛，就好像没有被捕获到过一样
    - 取throwed再throw是在当前位置重新将异常抛出

    需要特别注意：
        rethrow重新抛出的只是将此对象包含的异常和traceback重新打包，而不是重新抛出this本身
    */
    public void rethrow()

    /*
    返回包含的异常
    */
    public T throwed()

    /*
    返回traceback，是一个格式化好的字符串，只用于日志打印和人工排错
    */
    public String traceback()
}
```

### **直接捕获**

Larva提供了`handle_exc`内建函数和相关内建语法元素来实现直接捕获，使用方式可参考‘3.5.程序执行’，这里只对相关语法元素进行说明

```
/*
异常处理接口，用于handle_exc
*/
public interface ExcHandler<T>
{
    /*
    处理捕获到的异常
    */
    public void handle(Catched<T> catched_exc);
}

/*
捕获指定类型T的异常并调用handler的handle方法进行处理，需要和defer搭配使用，规则：
- 若无异常则不做操作（即也不会调用handler.handle(nil)，换句话说handle方法的参数不会是nil）
- 若有异常但不能转为T类型，则继续上抛，就像没有捕获过一样
- 若成功捕获到类型T的异常：
    - 若handler为nil，则忽略捕获的异常，相当于handler.handle是一个空方法
    - 否则调用handler.handle(catched_exc)处理异常
*/
public void handle_exc<T>(ExcHandler<T> handler)

/*
忽略所有异常，用于defer中，相当于handle_exc<Throwable>(nil)的简写
*/
public void ignore_exc()
```

### **返回值捕获**

Larva提供了另一个内建函数，可以使得异常像错误返回一样被处理：
```
/*
调用输入的Callable对象的call方法，规则：
- 如果callee的执行中没有出现异常，则返回nil
- 如果callee的执行中出现异常：
    - 若异常可以转为类型T，则以返回值的形式返回捕获的Catched<T>对象实例
    - 否则将异常上抛
*/
public Catched<T> call_and_catch<T>(Callable callee)
```

具体使用规则也可参考‘3.5.程序执行’

## **内建异常**

针对一些通用、常用的异常，Larva将其加入到了`__builtins`模块中作为内建类，方便开发者直接使用，这也有助于统一代码行为和风格，
本小节对这些异常类进行说明

注：由于异常类都是实现了`Throwable`接口，因此以下说明中，如异常的public方法只有一个`str`，则不列出代码
（这意味着构造函数也不是public的，使用者不可直接创建）

* `EmptyException`

    `EmptyException`是一个空异常类型，其`str`方法永远返回空串，作为异常本身来说，它并没有什么使用上的意义，一般是用于简单实现其他异常类型

    例如：
    ```
    class MyExc
    {
        EmptyException e usemethod;
    }
    ```
    这适用于使用者不关心异常的信息，而只关注异常类型的情形，例如通过`MyExc`来指示一些特别的错误，而不需要包含任何信息

* `TypeAssertionError`

    应用于内建函数`assert_type<T>`，当类型断言失败时抛出，`str`方法返回具体的类型转换失败信息

* `IndexError`

    索引越界错误，这里的索引是指下标，即在下标越界时抛出，或类似下标的语法元素越界时抛出

    ```
    public class IndexError
    {
        public IndexError(long i)

        public String str()
    }
    ```

    todo

    `throw_on_index_error`

* `RangeError`

    `throw_on_range_error`

* `KeyError`

* `ValueError`

* `GoError`

* `GoErrorPorter`
